File: clients/apple/Shared/Model/MeetingConfig.swift
import Foundation

struct MeetingConfig: Codable, Equatable {
    var signalingURLText: String
    var fallbackURLText: String
    var roomID: String
    var participantID: String? = nil
    var participantName: String
    var walletIdentity: String? = nil
    var requireSignedModeration: Bool = true
    var requirePaymentSettlement: Bool = false
    var preferWebFallbackOnPolicyFailure: Bool = true
    var supportsHDRCapture: Bool? = nil
    var supportsHDRRender: Bool? = nil

    static let `default` = MeetingConfig(
        signalingURLText: "ws://127.0.0.1:9000",
        fallbackURLText: "https://127.0.0.1:8080",
        roomID: "daily-standup",
        participantID: nil,
        participantName: "Alice",
        walletIdentity: "nexus://wallet/alice",
        requireSignedModeration: true,
        requirePaymentSettlement: false,
        preferWebFallbackOnPolicyFailure: true,
        supportsHDRCapture: nil,
        supportsHDRRender: nil
    )

    var signalingURL: URL? {
        normalizedURL(signalingURLText)
    }

    var fallbackURL: URL? {
        normalizedURL(fallbackURLText)
    }

    var isJoinable: Bool {
        signalingURL != nil && !roomID.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    private func normalizedURL(_ raw: String) -> URL? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, let url = URL(string: trimmed), let scheme = url.scheme else {
            return nil
        }
        let allowed = ["ws", "wss", "http", "https"]
        guard allowed.contains(scheme.lowercased()) else {
            return nil
        }
        return url
    }
}

File: clients/android/app/src/main/java/io/sora/kaigi/android/MeetingConfig.kt
package io.sora.kaigi.android

import java.net.URI

data class MeetingConfig(
    val signalingUrl: String = "ws://10.0.2.2:9000",
    val fallbackUrl: String = "https://10.0.2.2:8080",
    val roomId: String = "daily-standup",
    val participant: String = "Alice",
    val participantId: String? = null,
    val walletIdentity: String? = "nexus://wallet/alice",
    val requireSignedModeration: Boolean = true,
    val requirePaymentSettlement: Boolean = false,
    val preferWebFallbackOnPolicyFailure: Boolean = true,
    val supportsHdrCapture: Boolean? = null,
    val supportsHdrRender: Boolean? = null
) {
    fun signalingUriOrNull(): URI? = normalizedUri(signalingUrl)

    fun fallbackUriOrNull(): URI? = normalizedUri(fallbackUrl)

    fun isJoinable(): Boolean = signalingUriOrNull() != null && roomId.trim().isNotEmpty()

    private fun normalizedUri(raw: String): URI? {
        val trimmed = raw.trim()
        if (trimmed.isEmpty()) return null
        val uri = try {
            URI(trimmed)
        } catch (_: IllegalArgumentException) {
            return null
        }
        val scheme = uri.scheme?.lowercase() ?: return null
        if (scheme !in setOf("ws", "wss", "http", "https")) return null
        return uri
    }
}
